File: expAddOperator.py
# Time Complexity : O(2^N)
# Space Complexity : O(2^N)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def __init__(self):
        self.result = []
    def addOperators(self, num: str, target: int) -> List[str]:
        self.helper(num,target,[],0,0,0)
        return self.result


    def helper(self,num,target,path,index,calc,tail):
        if index == len(num):
            if calc == target:
                self.result.append("".join(path))
            return
        for i in range(index,len(num)):
            if num[index] == "0" and index != i:
                break

            curr = int(num[index:i+1])
            if index == 0:
                path.append(num[index:i+1])
                self.helper(num,target,path,i+1,curr,curr)
                path.pop()
            else:
                #+ 
                path.append("+")
                path.append(num[index:i+1])
                self.helper(num,target,path,i+1,calc+curr,curr)
                path.pop()
                path.pop()
                #-
                path.append("-")
                path.append(num[index:i+1])
                self.helper(num,target,path,i+1,calc-curr,-curr)
                path.pop()
                path.pop()
                #* 
                path.append("*")
                path.append(num[index:i+1])
                self.helper(num,target,path,i+1,calc-tail + tail*curr,tail*curr)
                path.pop()
                path.pop()

File: ExpAddOperator.java
//TC = N*3^N
//SC = O(N)

class Solution{
    List<String> result;
    public List<String> addOperators(String num, int target) {
        result = new ArrayList<>();
        helper(num, target, 0, 0, 0, "");
        return result;
    }
    private void helper(String num, int target, int pivote, long calc, long tail, String path){
        //base
        if(pivote == num.length()){
            if(calc == target){
                result.add(path.toString());
            }
            return;
        }

        //logic
        for(int i = pivote; i<num.length(); i++){
            if(num.charAt(pivote) == '0' && i!= pivote) continue;
            long curr = Long.parseLong(num.substring(pivote, i+1));
            //make conbibations
            if(pivote == 0){
                helper(num, target, i+1, curr, curr, path+curr);
            }else{
                //+
                //-
                //*
                helper(num, target, i+1, calc+curr, curr, path+ "+"+curr);
                helper(num, target, i+1, calc-curr, curr, path+ "-"+curr);
                helper(num, target, i+1, calc - tail + tail * curr, tail*curr, path+"*"+curr);
            }
        }

  
    }
}